### PR TITLE
MISC: Cargo.toml does not properly open as project in IDEA #2835

### LINKED ISSUE #2835: Cargo.toml does not properly open as project in IDEA
## Environment

* **Intellij-Rust plugin version:**
* **Rust toolchain version:** cargo 1.28.0
* **IDE name and version:** IntelliJ IDEA 2018.3 EAP
* **Operating system:** Linux

## Problem description

When I try to open project from file `Cargo.toml`, a new project is really created, but it shows only this file (`Cargo.toml`) instead of the complete file structure.

After closing the project and reopening it again, IDEA reports that the project does not exist; and it is probably right, because the path stored contains `Cargo.toml` instead of the last path element.

### LINKED ISSUE #2949: After opening a cargo project, it should be configured automatically
Currently, when I open an existing project (pointing to a directory; as opposed to a Cargo.toml #2835), source folders, libraries etc. are not configured until I explicitly open Rust files or Cargo project view.

Expected: upon opening, the project is configured by the IDE automatically

### PR DESCRIPTION
**Pull request by amakeev**
Fixes #2835

### CODE REVIEW

**Comment by vlad20012** [comment]
@Undin Should we do the same with Xargo.toml?

File: src/main/kotlin/org/rust/cargo/project/CargoProjectOpenProcessor.kt [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+import org.rust.cargo.icons.CargoIcons
+import org.rust.cargo.project.model.guessAndSetupRustProject
+import javax.swing.Icon
+
+class CargoProjectOpenProcessor : ProjectOpenProcessor() {
+    override fun getIcon(): Icon? = CargoIcons.ICON
+    override fun getName(): String = "Cargo"
+
+    override fun canOpenProject(file: VirtualFile?): Boolean = isCargoToml(file)
+    private fun isCargoToml(file: VirtualFile?) = FileUtil.namesEqual(file?.name, CargoConstants.MANIFEST_FILE)
```
Comment (Undin) [inline src/main/kotlin/org/rust/cargo/project/CargoProjectOpenProcessor.kt]: Maybe inline this method? Currently, we use it only in one place

Comment (Undin) [reply src/main/kotlin/org/rust/cargo/project/CargoProjectOpenProcessor.kt]: But if you think that it really simplify code, let's specify return type of `isCargoToml` explicitly

File: src/main/kotlin/org/rust/cargo/project/CargoProjectOpenProcessor.kt [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+    override fun doOpenProject(virtualFile: VirtualFile, projectToClose: Project?, forceNewFrame: Boolean): Project? {
+        val basedir = if (virtualFile.isDirectory) virtualFile else virtualFile.parent
+
+        val project = PlatformProjectOpenProcessor.getInstance().doOpenProject(basedir, projectToClose, forceNewFrame)
+        project?.let { StartupManager.getInstance(it).runWhenProjectIsInitialized { guessAndSetupRustProject(it) } }
+
+        return project
```
Comment (Undin) [inline src/main/kotlin/org/rust/cargo/project/CargoProjectOpenProcessor.kt]: Maybe it's a little readable?
```kotlin
val project = PlatformProjectOpenProcessor.getInstance()
    .doOpenProject(basedir, projectToClose, forceNewFrame) ?: return null
StartupManager.getInstance(project).runWhenProjectIsInitialized { guessAndSetupRustProject(project) }

return project
```

**Comment by Undin** [comment]
About `Xargo.toml` - as I understand `Xargo.toml` doesn't replace `Cargo.toml`, it just adds some specific properties for `xargo`, so I think we can take into account only `Cargo.toml` for now and implement xargo specific behavior in the future if we really need it

File: src/main/kotlin/org/rust/cargo/project/CargoProjectOpenProcessor.kt (hunk 1 above)
Comment (amakeev) [reply src/main/kotlin/org/rust/cargo/project/CargoProjectOpenProcessor.kt]: inlined

File: src/main/kotlin/org/rust/cargo/project/CargoProjectOpenProcessor.kt (hunk 2 above)
Comment (amakeev) [reply src/main/kotlin/org/rust/cargo/project/CargoProjectOpenProcessor.kt]: simplified

**Comment by Undin** [comment]
bors r-
Let's merge it after release

**Comment by Undin** [comment]
bors r=amakeev

**Pull request merged**